### PR TITLE
Keep original literal tokens in AST

### PR DESCRIPTION
**Pull request by petrochenkov**
The original literal tokens (`token::Lit`) are kept in AST until lowering to HIR.

The tokens are kept together with their lowered "semantic" representation (`ast::LitKind`), so the size of `ast::Lit` is increased (this also increases the size of meta-item structs used for processing built-in attributes).
However, the size of `ast::Expr` stays the same.

The intent is to remove the "semantic" representation from AST eventually and keep literals as tokens until lowering to HIR (at least), and I'm going to work on that, but it would be good to land this sooner to unblock progress on the [lexer refactoring](https://github.com/rust-lang/rust/pull/59706).

Fixes a part of https://github.com/rust-lang/rust/issues/43081 (literal tokens that are passed to proc macros are always precise, including hexadecimal numbers, strings with their original escaping, etc)
Fixes a part of https://github.com/rust-lang/rust/issues/60495 (everything except for proc macro API doesn't need escaping anymore)
This also allows to eliminate a certain hack from the lexer (https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/pretty-printing.20comments/near/165005357).

cc @matklad

### CODE REVIEW

File: src/libsyntax/ast.rs [hunk 1]
```diff
@@ -1354,6 +1355,8 @@ pub enum StrStyle {
 #[derive(Clone, RustcEncodable, RustcDecodable, Debug, Hash, PartialEq)]
 pub struct Lit {
     pub node: LitKind,
+    pub token: token::Lit,
```
Comment (matklad) [inline src/libsyntax/ast.rs]: Let's maybe add a comment here explaining why we need both `LitKind` and `token::Lit` and that we eventually plan to phase out `LitKind`. The problem with such a big incremental refactoring is that, once you are half-way through it, it's easy to forget which way you were going :D Explaining the intention in the comment should help.

I've seen a cute hack at JetBrains for this: when `X` is going to be replaced with `Y` across codebase, first `Y` is added alongside `X`. Then, `X` is marked as deprecated, and all the call-sites are marked with `#[allow(deprecated)]` (IDE helps a lot here). At this stage, new usages of `X` are not added, because they trigger deprecation warning, but old usages continue to work, without causing warnings. Then old usages are incrementally removed.

File: src/libsyntax/print/pprust.rs [hunk 2]
```diff
@@ -180,6 +162,31 @@ fn binop_to_string(op: BinOpToken) -> &'static str {
     }
 }
 
+fn literal_to_string(lit: token::Lit, suffix: Option<ast::Name>) -> String {
+    let mut out = match lit {
+        token::Byte(b)           => format!("b'{}'", b),
+        token::Char(c)           => format!("'{}'", c),
+        token::Err(c)            => format!("'{}'", c),
```
Comment (matklad) [inline src/libsyntax/print/pprust.rs]: Hm, I think we no longer use the `Err` token? Perhaps it makes sense to purge it in this PR as well?

I can imagine that, if we start using `Err` again, this `format!("'{}'", c),` might break (in a sense that it'll produce more than one token)

**Comment by petrochenkov** [comment]
@rust-timer build 11c353b37a551ef98c897ee4ac2235adbd603669

File: src/libsyntax/ast.rs (hunk 1 above)
Comment (petrochenkov) [reply src/libsyntax/ast.rs]: This depends on how much I manage to refactor today, perhaps I'll remove `node` entirely, then the deprecation will be unnecessary, otherwise I'll add the relevant comments.

File: src/libsyntax/print/pprust.rs (hunk 2 above)
Comment (petrochenkov) [reply src/libsyntax/print/pprust.rs]: `token::Err` is needed at least as a token counterpart to `LitKind::Err` (similarly to `token::Bool`).

Beside that, I have some plans for it.
It would be good for error recovery to have an "error token" that proc macros can accept and produce (AST based macros use `DummyResult::any`/`DummyResult::expr` for this to great success, see e.g. https://github.com/rust-lang/rust/pull/57205 and the issue it fixes).
A literal that's lexically invalid, but can survive through stringification would probably be a good candidate for such token.
(This needs some thought, so I'm unlikely to touch it in this PR.)

**Comment by petrochenkov** [comment]
Perf looks acceptable.

**Comment by petrochenkov** [comment]
Updated with two commits (one with refactoring and another moving code around).
This should be ready for review, I don't plan to do more refactoring in this PR (beside addressing https://github.com/rust-lang/rust/pull/60679#discussion_r282635984), except that the last commit crashes libcore build (?!) and I didn't yet figured out why.

@bors delegate=matklad

File: src/libsyntax/parse/literal.rs [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+        match $opt_diag {
+            Some(($span, $diag)) => { $($body)* }
+            None => return None,
+        }
+    }
+}
+
+impl LitKind {
+    /// Converts literal token with a suffix into a semantic literal.
+    /// Works speculatively and may return `None` is diagnostic handler is not passed.
```
[the remark this answers is not in the document]
Comment (matklad) [reply src/libsyntax/parse/literal.rs]: or maybe just drop the comment, as it duplicates the one on the more visible `crate fn`

File: src/libsyntax/parse/literal.rs [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+
+impl Lit {
+    /// Converts literal token with a suffix into an AST literal.
+    /// Works speculatively and may return `None` is diagnostic handler is not passed.
+    /// If diagnostic handler is passed, may return `Some`,
+    /// possibly after reporting non-fatal errors and recovery, or `None` for irrecoverable errors.
+    crate fn from_token(
+        token: &token::Token,
+        span: Span,
+        diag: Option<(Span, &Handler)>,
```
Comment (matklad) [inline src/libsyntax/parse/literal.rs]: `diag: Option` API feels awkward. From my reading of the code, the following is true

* `diag` is used **only** to report invalid suffix errors
* the fist call site is attrs/mod, where diag is always `None`
* the second call site is in this module, where diag is always `Some`

Perhaps we could remove `diag` argument altogether, and instead add an `fn validate_suffix(literal, suffix, span, handler)` function which we call from the parser? 

On the higher level, perhaps the problem is that attrs/mod doesn't have a diagnostic handler? It's not obvious why is it ok to ignore diagnostics there...

Comment (petrochenkov) [reply src/libsyntax/parse/literal.rs]: >On the higher level, perhaps the problem is that attrs/mod doesn't have a diagnostic handler?

The historical reason, AFAIK, is that meta-items can be parsed during command line parsing or pretty-printing and `Parser` (and `ParseSess` maybe) are unavailable at that time.
Perhaps things have changed because I see that parser/parse-session are created e.g. in `fn parse_cfgspecs` now.

If it's possible to thread parse session through all the meta-item parsing code that would be great, because `attrs/mod.rs` contains a lot of code that duplicates normal parsing.
That's a pretty large refactoring though, `my_attr.meta()` is used during processing of pretty much every built-in attribute and it doesn't use parse session.

>It's not obvious why is it ok to ignore diagnostics there...

Meta-item parsing works speculatively and if the result is `None` it reports one big "ill-formed meta item" error.

Comment (petrochenkov) [reply src/libsyntax/parse/literal.rs]: >diag is used only to report invalid suffix errors

There's also the "int literal is too large" error for literals not fitting into `u128`.

>Perhaps we could remove diag argument altogether, and instead add an fn validate_suffix(literal, suffix, span, handler) function which we call from the parser?

The suffix checking is somewhat intertwined with lowering `(token::Lit, Suffix) -> ast::LitKind` because lowering can turn integer literal token with suffix like `1f32` into floating point AST literals (see `fn integer_lit/float_lit/filtered_float_lit`), so I just kept those functions alone and didn't change them.
Right now we need the lowering part (as opposed to the validation part) anyway, because AST still contains `LitKind`, perhaps we can reconsider and split lowering and validation later after lowering only happens during HIR construction and after attr/mod has access to `Handler` (if possible).

Comment (matklad) [reply src/libsyntax/parse/literal.rs]: Yeah, makes sense to postpone this. Another alternative would be to change from `diag` input to `Result<LitKind, Error>` and feed `Error` into `Handler` in the parser and ignore in meta items, like unescaping works, but I imagine this won't be really better, given that the code in question is in the single module anyway

File: src/libsyntax/parse/literal.rs [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+                let diag = Some((span, &self.sess.span_diagnostic));
+                if let Some(lit) = Lit::from_token(&token, span, diag) {
+                    self.bump();
+                    self.bump();
+                    return Ok(lit);
+                }
+            }
+        }
+
+        Err(self.span_fatal(self.span, &format!("unexpected token: {}", self.this_token_descr())))
```
Comment (petrochenkov) [inline src/libsyntax/parse/literal.rs]: Drive-by fix: `self.prev_span` -> `self.span`, UI tests changed due to this.

**Comment by petrochenkov** [comment]
Updated.

**Comment by matklad** [comment]
@bors r+

Thanks!

**Comment by matklad** [comment]
Failure seems spurious?

**Comment by petrochenkov** [comment]
Yes, looks spurious.
@bors retry
